database/migrations/000093_profile_status_trigger.up.sql

```diff
@@ -0,0 +1,203 @@
+  -- queries hit the same rows, so they'll likely hit the cache.
+
+  -- These queries join on the latest_evaluation_statuses table to ensure that
+  -- we exclude historical statuses.
+
+  SELECT EXISTS (
+       SELECT 1 FROM latest_evaluation_statuses les
+       INNER JOIN evaluation_statuses es ON es.id = les.evaluation_history_id
+       WHERE les.profile_id = NEW.profile_id
+         AND es.status = 'error'
+  ) INTO v_other_error;
+
+  SELECT EXISTS (
+      SELECT 1 FROM latest_evaluation_statuses les
+      INNER JOIN evaluation_statuses es ON es.id = les.evaluation_history_id
+      WHERE les.profile_id = NEW.profile_id
+        AND es.status = 'failure'
+  ) INTO v_other_failed;
+
+  SELECT EXISTS (
+      SELECT 1 FROM latest_evaluation_statuses les
+      INNER JOIN evaluation_statuses es ON es.id = les.evaluation_history_id
+      WHERE les.profile_id = NEW.profile_id
+        AND es.status = 'success'
+  ) INTO v_other_success;
+
+  SELECT EXISTS (
+      SELECT 1 FROM latest_evaluation_statuses les
+      INNER JOIN evaluation_statuses es ON es.id = les.evaluation_history_id
+      WHERE les.profile_id = NEW.profile_id
+        AND es.status = 'skipped'
+  ) INTO v_other_skipped;
+
+  SELECT NOT EXISTS (
+      SELECT 1 FROM latest_evaluation_statuses les
+      INNER JOIN evaluation_statuses es ON es.id = les.evaluation_history_id
+      WHERE les.profile_id = NEW.profile_id
+  ) INTO v_pending;
+
+  CASE
+      -- A single rule in error state means policy is in error state
+      WHEN v_new_status = 'error' THEN
+          v_status := 'error';
+
+      -- No rule in error state and at least one rule in failure state
+      -- means policy is in error state
+      WHEN v_new_status = 'failure' AND v_other_error THEN
+          v_status := 'error';
+      WHEN v_new_status = 'failure' THEN
+          v_status := 'failure';
+
+      -- No rule in error or failure state and at least one rule in
+      -- success state means policy is in success state
+      WHEN v_new_status = 'success' AND v_other_error THEN
+          v_status := 'error';
+      WHEN v_new_status = 'success' AND v_other_failed THEN
+          v_status := 'failure';
+      WHEN v_new_status = 'success' THEN
+          v_status := 'success';
+
+      -- No rule in error, failure, or success state and at least one
+      -- rule in skipped state means policy is in skipped state
+      WHEN v_new_status = 'skipped' AND v_other_error THEN
+          v_status := 'error';
+      WHEN v_new_status = 'skipped' AND v_other_failed THEN
+          v_status := 'failure';
+      WHEN v_new_status = 'skipped' AND v_other_success THEN
+          v_status := 'success';
+      WHEN v_new_status = 'skipped' THEN
+          v_status := 'skipped';
+
+    -- This should never happen, if yes, make it visible
+    ELSE
+      v_status := 'error';
+      RAISE WARNING 'default case should not happen';
+  END CASE;
+
+  -- This turned out to be very useful during debugging
```

Comment:
This comment is copy-pasted from the previous trigger function, but I can testify that it was in fact _very useful_ during debugging.